tap_pendo/streams.py

```diff
@@ -600,19 +600,68 @@ def sync(self, state, start_date=None, key_id=None):
         # Set lookback window
         lookback = bookmark_dttm - timedelta(
             days=self.lookback_window())
-        ts = int(lookback.timestamp()) * 1000
 
+        # get events data
+        events = self.get_events(lookback, state, bookmark_dttm)
+        return (self.stream, events)
+
+    def get_events(self, window_start_date, state, bookmark_dttm):
+        # initialize start date as max bookmark
+        max_bookmark = bookmark_dttm
         # Get period type from config and make request for event's data
         period = self.config.get('period')
-        body = self.get_body(period, ts)
-        events = self.request(self.name, json=body) or []
-        update_currently_syncing(state, None)
-        return self.stream, events
+        # date format to filter
+        date = "date({}, {}, {})"
+        while True:
+            update_currently_syncing(state, self.name)
+
+            # get year, month and day from the start date
+            # if the start date is '2021-01-01' and 'events_date_window' is 25 days
+            # then the window end date will be '2021-01-26'
+            start_date, end_date = round_times(window_start_date, window_start_date + timedelta(days=int(self.config.get('events_date_window', 30))))
+
+            # create start filter
+            start = date.format(start_date.year, start_date.month, start_date.day)
+            # create end filter
+            end = date.format(end_date.year, end_date.month, end_date.day)
+
+            body = self.get_body(period, start, end)
+            events = self.request(self.name, json=body) or []
+            event = None
+
+            # loop over every event and yield
+            for event in events:
+                yield event
+
+            # as we are fetching the data in sorted manner (ascending),
+            # the last event will contain the highest bookmark value
+            if event:
+                replication_value = event.get(humps.decamelize(self.replication_key))
+                bookmark_value = strptime_to_utc(strftime(datetime.fromtimestamp(replication_value / 1000, timezone.utc)))
+                max_bookmark = max(bookmark_value, max_bookmark)
+                self.update_bookmark(state, self.name, strftime(max_bookmark), self.replication_key)
+
+            # for 'dayRange' the data is aggregated by day,
+            # hence adding 1 day more in the start date ie.
+            # Old Start date = date(2021, 1, 1)
+            # Old End date = date(2021, 1, 31)
+            # New Start date = date(2021, 2, 1)
+            # New End date = date(2021, 3, 3)
+            if self.period == 'dayRange':
+                window_start_date = end_date + timedelta(days=1)
+            else:
+                window_start_date = end_date
```

Comment:
@harshpatel4crest Can you please add a comment on the behavior that will be observed for hourRange?

Reply:
Added comment related to hourRange in the code.